### PR TITLE
support noaligned silu_and_mul

### PR DESCRIPTION
**Pull request by grimoire**
- support silu_and_mul with any input shape.
- optimize prefill attn kernel for deepseekv2

### CODE REVIEW

File: lmdeploy/pytorch/kernels/cuda/activation.py [hunk 1]
```diff
@@ -49,6 +49,42 @@ def _silu_and_mul_kernel(
         out_ptrs += BLOCK_SIZE_N * stride_on
 
 
+@triton.jit
+def _silu_and_mul_no_align_kernel(
```
Comment (RunningLeon) [inline lmdeploy/pytorch/kernels/cuda/activation.py:53]: failed with `CUDA_VISIBLE_DEVICES=0,1,2,3 lmdeploy chat ./DeepSeek-V2-Lite-Chat --tp 4`
```
loc("/home/root/lmdeploy/lmdeploy/pytorch/kernels/cuda/activation.py":33:26): error: Number of elements must be power-of-two, but %6 = "tt.make_range"() <{end = 352 : i32, start = 0 : i32}> : () -> tensor<352xi32> doesn't follow the rule (352) elements
2024-09-24 18:10:41,394 - lmdeploy - ERROR - model_agent.py:489 - Rank[2] failed.
loc(loc("/home/ma"/home/maningsningheng/sheng/llmdepmdeploloy/y/lmdeplmlodeploy/y/pytorpytorcch/khernels/cu/kda/aernelcs/cutivatida/acon.py":tivat33:ion.p26)y: error: Number of elements must be power-of-two, but %6 = "tt.make_range"() <{end = 352 : i32, start = 0 : i32}> : () -> tensor<352xi32> doesn't follow the rule (352)":33 elements
:26): error: Number of elements must be power-of-two, but %6 = "tt.make_range"() <{end = 352 : i32, start = 0 : i32}> : () -> tensor<352xi32> doesn't follow the rule (352) elements
2024-09-24 18:10:41,395 - lmdeploy - ERROR - model_agent.py:489 - Rank[3] failed.
2024-09-24 18:10:41,395 - lmdeploy - ERROR - model_agent.py:489 - Rank[1] failed.
```

**Review by AllentDan — APPROVED** [review]
LGTM

File: lmdeploy/pytorch/kernels/cuda/activation.py [hunk 2]
```diff
@@ -59,19 +95,32 @@ def silu_and_mul(gate_up: torch.Tensor, out: torch.Tensor = None):
...
+                                   N,
+                                   stride_gum=gate_up.stride(0),
+                                   stride_gun=gate_up.stride(1),
+                                   stride_om=out.stride(0),
+                                   stride_on=out.stride(1),
+                                   BLOCK_SIZE_N=BLOCK_SIZE_N,
+                                   num_warps=num_warps,
+                                   num_stages=num_stages)
+    else:
+        _silu_and_mul_no_align_kernel[grid](gate_up,
```
Comment (AllentDan) [inline lmdeploy/pytorch/kernels/cuda/activation.py:115]: may update unite test for no aligned case

File: lmdeploy/pytorch/models/minicpm3.py [hunk 3]
```diff
@@ -392,7 +394,7 @@ def get_input_embeddings(self):
         return self.embed_tokens
 
 
-class MiniCPM3ForCausalLM(nn.Module):
+class MiniCPM3ForCausalLM(nn.Module, CudaGraphMixin):
```
Comment (RunningLeon) [inline lmdeploy/pytorch/models/minicpm3.py:397]: failed with `CUDA_VISIBLE_DEVICES=2,3 lmdeploy chat /mnt/140/MiniCPM3-4B --tp 2`
```
  File "/home/root/lmdeploy/lmdeploy/pytorch/models/minicpm3.py", line 139, in forward
    kv = (self.kv_b_proj(self.kv_a_layernorm(compressed_kv)).view(
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: shape '[1, 32, 20, 128]' is invalid for input of size 163840
```

**Review by RunningLeon — APPROVED** [review]
LGTM

**Pull request merged**